### DISCUSSION TITLE
support for range formatting

### DISCUSSION THREAD

**Discussion by IanRPage** [rust-lang/rust-analyzer#17899] [General]
Hello anyone reading this, I was wondering if there is any plans for further development of support range formatting. In the Configuration section of the manual there's a `rust-analyzer.rustfmt.rangeFormatting.enable` config that enables rustfmt's unstable range formatting for textDocument/rangeFormatting on nightly build. I've enabled this configuration.

I'm not sure if it's relevant to this discussion, but I'm using Helix and am trying to format a selection (using `=` key bound to `format_selections`). When I'm running RA stable, I get the error `No configured language server supports range formatting` when trying to format selection, and when I'm using RA nightly (64a1405) the error doesn't pop up but the code I selected doesn't get formatted.

Also, I've enabled this in my helix/languages.toml file. But is there a separate place I need to enable this for RA itself? I installed the RA nightly binary, if that's of importance.

By the way, RA is awesome so thank you to all the people who are contributing to it!

**Comment by lnicola** [comment]
I'm not sure about helix, but in VS Code I had to:

 - run `rustup +nightly component add rustfmt` (you also need to install the toolchain, of course)
 - enable `rust-analyzer.rustfmt.rangeFormatting.enable`
 - restart
 - set `rust-analyzer.rustfmt.extraArgs` to `["+nightly"]`

**Reply by IanRPage** [reply]
Okay, I'll give that a shot. Thanks

**Comment by Kyzzsa** [comment]
I’m also using Helix and just thought it would be nice to have a specific range formatting configuration, partially expanding complex code more than the default configuration, so I can easily do stuff like swapping function parameters, and finally collapse them to the default format on save.

**Reply by ChayimFriedman2** [reply]
I'm not sure I understand you, but swapping function parameters is not related to formatting. And like said above rust-analyzer already supports range formatting when rustfmt supports it.

**Reply by Kyzzsa** [reply]
@ChayimFriedman2
I mean I can have a range formatting config in which `max_width` is set to a very small value.
So I can format
``` rust
foo((bar + baz), baz, (foobar * foobar))
```
into
``` rust
foo(
    (bar + baz),
    baz,
    (foobar * foobar)
)
```
, then I can easily change the arguments' order by moving lines up and down, instead of selecting arguments and inserting them after a comma one by one.

**Reply by Kyzzsa** [reply]
Thanks for the reply btw.

**Reply by ChayimFriedman2** [reply]
That is unrelated to rust-analyzer. You should request rustfmt to have a different config for range formatting.

**Reply by lnicola** [reply]
Or it's a job for SSR.

**Comment by wazer24** [comment]
For range formatting with rust-analyzer you'll likely need to try nightly rustfmt as it's an unstable feature:

## Configuration

Make sure to install the nightly components:

```bash
rustup toolchain install nightly --component rustfmt
```

(if nightly hasn't already been installed just run `rustup +nightly component add rustfmt`)

Set this in your editor LSP config for rust-analyzer:

```json
"rust-analyzer.rustfmt.rangeFormatting.enable": true
```

Also, add `"+nightly"` to the `rust-analyzer.rustfmt.extraArgs`:

```json
"rust-analyzer.rustfmt.extraArgs": ["+nightly"]
```

Now you'll need to restart the language server and possibly your editor to make sure it takes effect.

## Format vs Refactor

Just a quick note that "formatting" only changes what your code looks like (e.g. The spacing, indentation and line lengths), if you wanted to swap the parameters of your function that's not considered a formatting action-it's refactoring. If that's what you're interested in then you'll need to look for some semantic code action (you might see this referred to as "assists" in rust-analyzer terminology) not range formatting.
